File: map.go
// map related things
package moo

import (
	"github.com/DrItanium/moo/cseries"
)

const (
	TicksPerSecond = 30
	TicksPerMinute = 60 * TicksPerSecond

	MapIndexBufferSize              = 8192
	MinimumSeparationFromWall       = WorldOne / 4
	MinimumSeparationFromProjectile = ((3 * WorldOne) / 4)

	TeleportingMidpoint = TicksPerSecond / 2
	TeleportingDuration = 2 * TeleportingMidpoint

	MaximumPolygonsPerMap  = cseries.Kilo
	MaximumSidesPerMap     = 4 * cseries.Kilo
	MaximumEndpointsPerMap = 8 * cseries.Kilo
	MaximumLinesPerMap     = 4 * cseries.Kilo
	MaximumLevelsPerMap    = 128

	LevelNameLength = 64 + 1
)
const (
	// damage types
	DamageExplosion = iota
	DamageElectricalStaff
	DamageProjectile
	DamageAbsorbed
	DamageFlame
	DamageHoundClaws
	DamageAlienProjectile
	DamageHulkSlap
	DamageCompilerBolt
	DamageFusionBolt
	DamageHunterBolt
	DamageFist
	DamageTeleporter
	DamageDefender
	DamageYetiClaws
	DamageYetiProjectile
	DamageCrushing
	DamageLava
	DamageSuffocation
	DamageGoo
	DamageEnergyDrain
	DamageOxygenDrain
	DamageHummerBolt
	DamageShotgunProjectile
)
const (
	// damage flags
	AlienDamage = 0x1

	MaximumSavedObjects = 384
)
const (
	// map object types
	SavedMonster     = iota // index is monster type
	SavedObject             // index is scenery type
	SavedItem               // index is item type
	SavedPlayer             // index is team bitfield
	SavedGoal               // index is goal number
	SavedSoundSource        // index is source type, facing is sound volume
)
const (
	MapObjectIsInvisible = 0x0001

	MaximumVerticesPerPolygon = 8

	MaximumObjectTypes = 64
)

type DamageDefinition struct {
	Type   int16
	Flags  int16
	Base   int16
	Random int16
	Scale  cseries.Fixed
}
type SideTextureDefinition struct {
	X0, Y0  WorldDistance
	Texture ShapeDescriptor
}

type ObjectLocation struct {
	P            WorldPoint3d
	PolygonIndex int16

	Yaw, Pitch Angle
	Flags      cseries.Word
}
type StaticData struct {
	EnviromentCode int16

	PhysicsModel     int16
	SongIndex        int16
	MissionFlags     int16
	EnvironmentFlags int16

	BallInPlay bool // true if there's a ball in play
	unused1    bool
	unused     [3]int16

	LevelName       string
	EntryPointFlags int32
}
type DynamicData struct {
	// ticks since the beginning of the game
	TickCount int32

	// the real seed is static in WORLD.C; must call set_random_seed()
	RandomSeed cseries.Word

	// this is stored in the dynamic_data so that it is valid across saves
	//GameInformation GameData

	PlayerCount         int16
	SpeakingPlayerIndex int16

	unused                                        int16
	PlatformCount                                 int16
	EndpointCount                                 int16
	LineCount                                     int16
	SideCount                                     int16
	PolygonCount                                  int16
	LightsourceCount                              int16
	MapIndexCount                                 int16
	AmbientSoundImageCount, RandomSoundImageCount int16

	//statistically unlikely to be valid

	ObjectCount     int16
	MonsterCount    int16
	ProjectileCount int16
	EffectCount     int16
	LightCount      int16

	DefaultAnnotationCount  int16
	PersonalAnnotationCount int16

	InitialObjectsCount int16

	GarbageObjectCount int16

	// used by MoveMonsters to decide who gets to generate paths, etc.
	LastMonsterIndexToGetTime, LastMonsterIndexToBuildPath int16

	// variables used by NewMonster to adjust for different difficulty levels
	NewMonsterManglerCookie, NewMonsterVanishingCookie int16

	// number of civilians killed by players; periodically decremented
	CiviliansKilledByPlayers int16

	// Used by the item placement stuff
	RandomMonstersLeft  [MaximumObjectTypes]int16
	CurrentMonsterCount [MaximumObjectTypes]int16
	RandomItemsLeft     [MaximumObjectTypes]int16
	CurrentItemCount    [MaximumObjectTypes]int16

	CurrentLevelNumber int16 // what level the user is currently exploring

	CurrentCivilianCausalties, CurrentCivilianCount int16
	TotalCivilianCausalties, TotalCivilianCount     int16

	GameBeacon      WorldPoint2d
	GamePlayerIndex int16
}

var StaticWorld *StaticData
var DynamicWorld *DynamicData

type ObjectData struct {
	Location WorldPoint3d
	Polygon  int16

	Facing Angle

	Shape ShapeDescriptor

	Sequence, Flags cseries.Word

	TransferMode, TransferPeriod int16

	TransferPhase   int16
	Permutation     int16
	NextObject      int16
	ParasiticObject int16

	SoundPitch cseries.Fixed
}

type EndpointData struct {
	Solid       bool
	Transparent bool
	Elevation   bool

	HighestAdjacentFloorHeight, LowestAdjacentCeilingHeight WorldDistance
	Vertex                                                  WorldPoint2d
	Transformed                                             WorldPoint2d

	SupportingPolygonIndex int16
}

type LineData struct {
	EndpointIndexes [2]int16

	Solid                                       bool
	Transparent                                 bool
	Landscape                                   bool
	Elevation                                   bool
	VariableElevation                           bool
	LineHasTransparentSide                      bool
	Length                                      WorldDistance
	HighestAdjacentFloor, LowestAdjacentCeiling WorldDistance

	ClockwisePolygonSideIndex, CounterClockwisePolygonSideIndex int16

	ClockwisePolygonOwner, CounterClockwisePolygonOwner int16
}

type SideExclusionZone struct {
	E0, E1, E2, E3 WorldPoint2d
}

type ControlPanelSideType int16

const (
	ControlPanelOxygenRefuel ControlPanelSideType = iota
	ControlPanelShieldRefuel
	ControlPanelDoubleShieldRefuel
	ControlPanelTripleShieldRefuel
	ControlPanelLightSwitch
	ControlPanelPlatformSwitch
	ControlPanelTagSwitch
	ControlPanelPatternBuffer
	ControlPanelComputerTerminal

	NumberOfControlPanels
)

type SideData struct {
	Type int16

	ControlPanelStatus        bool
	ControlPanel              bool
	RepairSwitch              bool
	DestructiveSwitch         bool
	LightedSwitch             bool
	CanBeDestroyed            bool
	CanOnlyBeHitByProjectiles bool

	EditorDirty bool // this is probably not relevant

	Primary     SideTextureDefinition
	Secondary   SideTextureDefinition
	Transparent SideTextureDefinition

	ExclusionZone SideExclusionZone

	ControlPanelType        ControlPanelSideType
	ControlPanelPermutation int16

	PrimaryTransferMode     int16
	SecondaryTransferMode   int16
	TransparentTransferMode int16

	PolygonIndex, LineIndex int16

	PrimaryLightsourceIndex     int16
	SecondaryLightsourceIndex   int16
	TransparentLightsourceIndex int16

	AmbientDelta cseries.Fixed
}

type PolygonType int16

const (
	NormalPolygon PolygonType = iota
	ItemImpassablePolygon
	MonsterImpassiblePolygon
	HillPolygon
	BasePolygon
	PlatformPolygon
	LightOnTriggerPolygon
	PlatformOnTriggerPolygon
	LightOffTriggerPolygon
	PlatformOffTriggerPolygon
	TeleporterPolygon
	ZoneBorderPolygon
	GoalPolygon
	VisibleMonsterTriggerPolygon
	InvisibleMonsterTriggerPolygon
	DualMonsterTriggerPolygon
	ItemTriggerPolygon
	MustBeExploredPolygon
	AutomaticExitPolygon
)

type HorizontalSurfaceData struct {
	Height                         WorldDistance
	LightSourceIndex               int16
	Texture                        ShapeDescriptor
	TransferMode, TransferModeData int16

	Origin WorldPoint2d
}

type PolygonData struct {
	Type        int16
	Flags       cseries.Word
	Permutation int16

	VertexCount     int16
	EndpointIndexes [MaximumVerticesPerPolygon]int16
	LineIndexes     [MaximumVerticesPerPolygon]int16

	FloorTexture, CeilingTexture                   ShapeDescriptor
	FloorHeight, CeilingHeight                     WorldDistance
	FloorLightsourceIndex, CeilingLightsourceIndex int16

	Area int32 // in WorldDistance^2 units

	FirstObject int16

	FirstExclusionZoneIndex int16
	LineExclusionZoneCount  int16
	PointExclusionZoneCount int16

	FloorTransferMode   int16
	CeilingTransferMode int16

	AdjacentPolygonIndexes [MaximumVerticesPerPolygon]int16

	FirstNeightborIndex int16
	NeighborCount       int16

	Center WorldPoint2d

	SideIndexes [MaximumVerticesPerPolygon]int16

	FloorOrigin, CeilingOrigin WorldPoint2d

	MediaIndex            int16
	MediaLightsourceIndex int16

	SoundSourceIndexes int16

	AmbientSoundImageIndex, RandomSoundImageIndex int16
}

type GameDifficultyLevel int16

const (
	DifficultyWuss GameDifficultyLevel = iota
	DifficultyEasy
	DifficultyNormal
	DifficultyMajorCarnage
	DifficultyTotalCarnage
	DifficultyCount
)

const MaximumObjectTypes = 64

const (
	// flags for object_frequency_definition
	ReappearsInRandomLocations = 0x0001
)

type ObjectFrequencyDefinition struct {
	Flags cseries.Word

	InitialCount int16 // number that initially appear. can be greater than maximum_count
	MinimumCount int16 // this number of objects will be maintained.
	MaximumCount int16 // can't exceed this, except at the beginning of the level.

	RandomCount  int16        // maximum random occurences of the object
	RandomChance cseries.Word // in (0, 65535]
}

type MissionFlagsDescription int16

const (
	// mission flags
	MissionNone          MissionFlagsDescription = 0x0000
	MissionExtermination                         = 0x0001
	MissionExploration                           = 0x0002
	MissionRetrieval                             = 0x0004
	MissionRepair                                = 0x0008
	MissionRescue                                = 0x0010
)

type EnvironmentFlagsDescription int16

const (
	/* environment flags */
	EnvironmentNormal     EnvironmentFlagsDescription = 0x0000
	EnvironmentVacuum                                 = 0x0001 // prevents certain weapons from working, player uses oxygen
	EnvironmentMagnetic                               = 0x0002 // motion sensor works poorly
	EnvironmentRebellion                              = 0x0004 // makes clients fight pfhor
	EnvironmentLowGravity                             = 0x0008 // low gravity

	EnvironmentNetwork      = 0x2000 // these two pseudo-environments are used to prevent items
	EnvironmentSinglePlayer = 0x4000 // from arriving in the items.c code.
)

type StaticData struct {
	EnvironmentCode int16

	PhysicsModel     int16
	SongIndex        int16
	MissionFlags     MissionFlagsDescription
	EnvironmentFlags EnvironmentFlagsDescription

	BallInPlay      bool
	LevelName       string
	EntryPointFlags int32
}
type GameOptions int16

const ( /* game options.. */
	MultiplayerGame             GameOptions = 0x0001 /* multi or single? */
	AmmoReplenishes                         = 0x0002 /* Does or doesn't */
	WeaponsReplenish                        = 0x0004 /* Weapons replenish? */
	SpecialsReplenish                       = 0x0008 /* Invisibility, Ammo? */
	MonstersReplenish                       = 0x0010 /* Monsters are lazarus.. */
	MotionSensorDoesNotWork                 = 0x0020 /* Motion sensor works */
	OverheadMapIsOmniscient                 = 0x0040 /* Only show teammates on overhead map */
	BurnItemsOnDeath                        = 0x0080 /* When you die, you lose everything but the initial crap.. */
	LiveNetworkStats                        = 0x0100
	GameHasKillLimit                        = 0x0200 /* Game ends when the kill limit is reached. */
	ForceUniqueTeams                        = 0x0400 /* every player must have a unique team */
	DyingIsPenalized                        = 0x0800 /* time penalty for dying */
	SuicideIsPenalized                      = 0x1000 /* time penalty for killing yourselves */
	OverheadMapShowsItems                   = 0x2000
	OverheadMapShowsMonsters                = 0x4000
	OverheadMapShowsProjectiles             = 0x8000
)

const (
	LevelUnfinished = iota
	LevelFinished
	LevelFailed
)

type GameType int16

// game types
const (
	GameOfKillMonsters    GameType = iota // single player & combative use this
	GameOfCooperativePlay                 // multiple players, working together
	GameOfCaptureTheFlag                  // A team game.
	GameOfKingOfTheHill
	GameOfKillManWithBall
	GameOfDefense
	GameOfRugby
	GameOfTag
	NumberOfGameTypes
)

//#define GET_GAME_TYPE() (dynamic_world->game_information.game_type)
//#define GET_GAME_OPTIONS() (dynamic_world->game_information.game_options)
//#define GET_GAME_PARAMETER(x) (dynamic_world->game_information.parameters[(x)])

type GameData struct {
	TimeRemaining     int32
	Type              GameType
	Options           GameOptions
	KillLimit         int16
	InitialRandomSeed int16
	Difficulty        GameDifficultyLevel
	// Parameters [2]int16 // use these later. for now memset to 0
}

type DynamicData struct {
	TickCount int32

	RandomSeed cseries.Word

	// this is stored in the DynamicData so that it is valid across saves
	GameInformation GameData

	PlayerCount         int16
	SpeakingPlayerIndex int16

	// Unused int16
	PlatformCount                                 int16
	EndpointCount                                 int16
	LineCount                                     int16
	SideCount                                     int16
	PolygonCount                                  int16
	LightsourceCount                              int16
	MapIndexCount                                 int16
	AmbientSoundImageCount, RandomSoundImageCount int16

	// statistically unlikely to be valid
	ObjectCount             int16
	MonsterCount            int16
	ProjectileCount         int16
	EffectCount             int16
	LightCount              int16
	DefaultAnnotationCount  int16
	PersonalAnnotationCount int16
	InitialObjectsCount     int16
	GarbageObjectCount      int16

	/* used by move_monsters() to decide who gets to generate paths, etc. */
	LastMonsterIndexToGetTime, LastMonsterIndexToBuildPath int16

	/* variables used by new_monster() to adjust for different difficulty levels */
	NewMonsterManglerCookie, NewMonsterVanishingCookie int16

	/* number of civilians killed by players; periodically decremented */
	CiviliansKilledByPlayers int16

	/* used by the item placement stuff */
	RandomMonstersLeft  [MaximumObjectTypes]int16
	CurrentMonsterCount [MaximumObjectTypes]int16
	RandomItemsLeft     [MaximumObjectTypes]int16
	CurrentItemCount    [MaximumObjectTypes]int16

	CurrentLevelNumber int16 // what level the user is currently exploring.

	CurrentCivilianCausalties, CurrentCivilianCount int16
	TotalCivilianCausalties, TotalCivilianCount     int16

	GameBeacon      WorldPoint2d
	GamePlayerIndex int16
}

var StaticWorld StaticData
var DynamicWorld DynamicData
var Objects []ObjectData

var MapPolygons []PolygonData
var MapSides []SideData
var MapLines []LineData
var MapEndpoints []EndpointData

var AmbientSoundImages []AmbientSoundImageData
var RandomSoundImages []RandomSoundImageData

var MapIndexes []short

var AutomapLines []byte
var AutomapPolygons []byte

var MapAnnotations []MapAnnotation
var SavedObjects []MapObject

var GameIsNetworked bool // true if this is a network game

//#define ADD_LINE_TO_AUTOMAP(i) (automap_lines[(i)>>3] |= (byte) 1<<((i)&0x07))
//#define LINE_IS_IN_AUTOMAP(i) ((automap_lines[(i)>>3]&((byte)1<<((i)&0x07)))?(TRUE):(FALSE))
//
//#define ADD_POLYGON_TO_AUTOMAP(i) (automap_polygons[(i)>>3] |= (byte) 1<<((i)&0x07))
//#define POLYGON_IS_IN_AUTOMAP(i) ((automap_polygons[(i)>>3]&((byte)1<<((i)&0x07)))?(TRUE):(FALSE))

File: map.fs
\ depends on world.fs

\ framerate information, yes this game is tied to framerate so 60fps causes
\ issues as the game will run twice as fast!
30 constant ticks-per-second
ticks-per-second 60 * constant ticks-per-minute

8192 constant map-index-buffer-size
world-one 4 / constant minimum-separation-from-wall
world-one *3/4 constant minimum-separation-from-projectile

ticks-per-second 2/ constant teleporting-midpoint
teleporting-midpoint 2* constant teleporting-duration

\ map limits
kilo constant max-polygons-per-map
kilo 4 * constant max-sides-per-map
kilo 8 * constant max-endpoints-per-map 
kilo 4 * constant max-lines-per-map
128 constant max-levels-per-map

64 1+ constant level-name-len

\ shape descriptor is included here... not sure why

{enum \ damage types
enum: DamageExplosion
enum: DamageElectricalStaff
enum: DamageProjectile
enum: DamageAbsorbed
enum: DamageFlame
enum: DamageHoundClaws
enum: DamageAlienProjectile
enum: DamageHulkSlap
enum: DamageCompilerBolt
enum: DamageFusionBolt
enum: DamageHunterBolt
enum: DamageFist
enum: DamageTeleporter
enum: DamageDefender
enum: DamageYetiClaws
enum: DamageYetiProjectile
enum: DamageCrushing
enum: DamageLava
enum: DamageSuffocation
enum: DamageGoo
enum: DamageEnergyDrain
enum: DamageOxygenDrain
enum: DamageHummerBolt
enum: DamageShotgunProjectile
enum}

\ damage flags
0x1 constant FlagAlienDamage \ will be decreased at lower difficulty levels

\ damage-definition struct
{struct
 field(int16): &damage-definition.type 
 field(int16): &damage-definition.flags
 field(int16): &damage-definition.base
 field(int16): &damage-definition.random
 field(int32): &damage-definition.scale
struct} constant sizeof(damage-definition)

384 constant max-saved-objects

{enum \ map object types
enum: SavedMonster \ .index is monster type 
enum: SavedObject \ scenery type
enum: SavedItem \ item type
enum: SavedPlayer \ team bitfield
enum: SavedGoal \ goal number
enum: SavedSoundSource \ source type with .facing being sound volume
enum}


\ map object flags
0x0001 constant MapObjectIsInvisibleFlag \ initially invisible 
0x0001 constant MapObjectIsPlatformSoundFlag 
0x0002 constant MapObjectIsHangingFromCeilingFlag \ used for calculating absolute .z coordinate
0x0004 constant MapObjectIsBlindFlag \ monster cannot activate by sight
0x0008 constant MapObjectIsDeafFlag \ monster cannot activate by sound
0x0010 constant MapObjectFloatsFlag \ used by sound sources caused by media 
0x0020 constant MapObjectIsNetworkOnlyFlag \ for items only
\ top four bits is activation bias for monsters
: bitset? ( value flag -- f ) and 0<> ;

: decode-activation-bias ( f -- n ) 12 u>> ;
: encode-activation-bias ( f -- n ) 12 u<< ;

{struct \ map object
 field(int16): &map-object.type 
 field(int16): &map-object.index
 field(int16): &map-object.facing
 field(int16): &map-object.polygon-index
 field(int16): &map-object.location
 sizeof(world-point3d) field: &map-object.location 
 field(int16): &map-object.flags
struct} constant sizeof(map-object)
\ sizeof(short) 4 * sizeof(word-point3d) + sizeof(word) + constant sizeof(map-object)
: map-object.flags@ ( adr -- f ) &map-object.flags + @ ;
: map-object.type@ ( adr -- f ) &map-object.type + @ ;
: map-object.flags! ( value adr -- f ) &map-object.flags + ! ;
: ?map-object.invisible ( adr -- f ) map-object.flags@ MapObjectIsInvisibleFlag bitset? ;
: ?map-object.platform-sound ( adr -- f ) map-object.flags@ MapObjectIsPlatformSoundFlag bitset? ;
: ?map-object.hanging-from-ceiling ( adr -- f ) map-object.flags@ MapObjectIsHangingFromCeilingFlag bitset? ;
: ?map-object.blind ( adr -- f ) map-object.flags@ MapObjectIsBlindFlag bitset? ;
: ?map-object.deaf ( adr -- f ) map-object.flags@ MapObjectIsDeafFlag bitset? ;
: ?map-object.floats ( adr -- f ) map-object.flags@ MapObjectFloatsFlag bitset? ;
: ?map-object.network-only ( adr -- f ) map-object.flags@ MapObjectIsNetworkOnlyFlag bitset? ;

: ?map-object.monster ( t -- f ) map-object.type@ SavedMonster = ;
: ?map-object.object ( t -- f ) map-object.type@ SavedObject = ;
: ?map-object.item ( t -- f ) map-object.type@ SavedItem = ;
: ?map-object.player ( t -- f ) map-object.type@ SavedPlayer = ;
: ?map-object.goal ( t -- f ) map-object.type@ SavedGoal = ;
: ?map-object.sound-source ( t -- f ) map-object.type@ SavedSoundSource = ;
: mo-index<-> ( -- ) &map-object.index <-> ;

mo-index<-> &map-object.sound-type
&map-object.facing <-> &map-object.sound-volume
mo-index<-> &map-object.scenery-type
mo-index<-> &map-object.monster-type
mo-index<-> &map-object.item-type
mo-index<-> &map-object.team-bitfield
mo-index<-> &map-object.goal-number
&map-object.flags <-> &map-object.monster-activation-bias

: map-object.monster-activation-bias@ ( adr -- f ) 
  &map-object.monster-activation-bias + @ 
  decode-activation-bias ;
: map-object.monster-activation-bias! ( v adr -- ) 
  &map-object.monster-activation-bias + dup @ 
  swap encode-activation-bias or swap ! ;
  
sizeof(world-point2d) constant sizeof(saved-map-pt)

\ line-data <-> saved-line
\ side-data <-> saved-side
\ polygon-data <-> saved-poly
\ map-annotation <-> saved-annotation
\ map-object <-> saved-object
\ static-data <-> saved-map-data

\ entry point types
0x01 constant EntryPointSinglePlayer
0x02 constant EntryPointMultiplayerCooperative
0x04 constant EntryPointMultiplayerCarnage
0x08 constant EntryPointCaptureTheFlag
0x10 constant EntryPointKingOfHill
0x20 constant EntryPointDefense
0x40 constant EntryPointRugby

{struct
 field(short): &entry-point.level-number
 level-name-len field: &entry-point.name 
struct} constant sizeof(entry-point)

32 constant max-player-start-name-len 

{struct \ player-start-data
 field(short): &player-start-data.team
 field(short): &player-start-data.identifier
 field(short): &player-start-data.color
 max-player-start-name-len 1+ field: &player-start-data.name 
struct} constant sizeof(player-start-data)

{struct \ directory-data
 field(short)           &directory-data.mission-flags
 field(short)           &directory-data.environment-flags
 field(long)            &directory-data.entry-point-flags
 level-name-len field:  &directory-data.level-name 
struct} constant sizeof(directory-data)

20 constant max-annotations-per-map
64 constant max-annotation-text-len
{struct \ map-annotation
    field(short): &map-annotation.type
    field(world-point2d): &map-annotation.location
    field(short): &map-annotation.polygon-index
    max-annotation-text-len field: text
struct} constant sizeof(map-annotation)

: get-next-map-annotation ( *count -- *ma ) ;

64 constant max-ambient-sound-images-per-map

{struct \ ambient-sound-image-data
\ non directional ambient component
  field(word): &ambient-sound-image-data.flags
  field(short): &ambient-sound-image-data.sound_index
  field(short): &ambient-sound-image-data.volume
  5 sizeof(short) * field: &ambient-sound-image-data.unused
struct} constant sizeof(ambient-sound-image-data)

64 constant max-random-sound-images-per-map

0x0001 constant SoundImageIsNonDirectionalFlag \ ignore direction

{struct \ random-sound-image-data
 field(word):             &random-sound-image-data.flags
 field(short):            &random-sound-image-data.sound-index
 field(short):            &random-sound-image-data.volume 
 field(short):            &random-sound-image-data.delta-volume 
 field(short):            &random-sound-image-data.period
 field(short):            &random-sound-image-data.delta-period
 field(angle):            &random-sound-image-data.direction
 field(angle):            &random-sound-image-data.delta-direction
 field(fixed):            &random-sound-image-data.pitch
 field(fixed):            &random-sound-image-data.delta-pitch
 field(short):            &random-sound-image-data.phase \ only used at runtime; initialize to none
 3 sizeof(short) * field: &random-sound-image-data.unused
struct} constant sizeof(random-sound-image-data)
\ the map contains a series of slots in which map objects are placed
384 constant max-objects-per-map

: ?flag-marked ( adr mask -- b ) swap word@ and (word) ;
: ?slot-used ( adr -- b ) 0x8000 ?flag-marked ;
: ?slot-free ( adr -- b ) ?slot-used invert (word) ;
: set-flag ( adr flag -- ) (word) over word@ or swap word! ;
: clear-flag ( adr flag -- ) invert (word) over word@ and swap word! ;
: mark-slot-as-free ( adr -- ) 0x8000 clear-flag ;
: mark-slot-as-used ( adr -- ) 0x8000 set-flag ;
: ?object-was-rendered ( adr -- f ) 0x4000 ?flag-marked ;
: mark-object-rendered ( adr -- ) 0x4000 set-flag ;
: mark-object-not-rendered ( adr -- ) 0x4000 clear-flag ;
\ map.h: line 238
;s \ always last in the file
